src/ccs/search_state.py

```diff
@@ -145,11 +196,72 @@ def match_step(key, value):
                                 poison(node)
                         # TODO and of course dually negative matches too
 
+        if activate_root:
+            keys = deque(self.dag.prop_node.constraints) + keys
+            activate(self.dag.prop_node)
+
         while keys:
             key = keys.popleft()
             assert len(key.values) < 2
             match_step(key.name, next(iter(key.values), None))
 
-        return Context(
-            self.dag, self.prop_accumulator, tallies, or_specificities, props, poisoned
+        return {
+            "tallies": tallies,
+            "or_specificities": or_specificities,
+            "props": props,
+            "poisoned": poisoned,
+        }
+
+    def get_single_property(self, prop: str) -> Property:
+        contenders = self.props.get(prop, None)
+        if contenders is None:
+            raise MissingPropertyError(f"Invalid property: {prop}")
+
+        properties = list(contenders.values)
+        if len(properties) == 0:
+            raise EmptyPropertyError(f"Property {prop} has no values")
+        if len(properties) > 1:
+            raise AmbiguousPropertyError(
+                f"Property {prop} has too many values: {properties}"
+            )
```

Comment:
Note: This is breaking compared to CCS 1.0. As far as I could see, we don't have access to source info (e.g. which file properties come from) at this level, which makes it hard to decide which property would be better in the case of a tie. Adding that tracing would help the debug printing as well, so it's probably worth having

Reply:
Yes on the tracing and origin tracking.

The resolution of conflicts based on file order has been mostly fine, but definitely not perfect, and I've thought it would be good to have a mode where conflicts are reported as an error, at the client's option. I'd rather keep the APIs consistent in the long run, but this is great for now.